sotodlib/hwp/g3thwp.py

```diff
@@ -1277,52 +1280,48 @@ def _hwp_angle_calculator(
-        for i in range(len(diff_split)):
-            _diff = diff_split[i]
-            # eliminate upper/lower _slit_width_lim
-            _diff_upperlim = np.percentile(
-                _diff, (1 - self._slit_width_lim) * 100)
-            _diff_lowerlim = np.percentile(_diff, self._slit_width_lim * 100)
-            __diff = _diff[np.where(
-                (_diff < _diff_upperlim) & (_diff > _diff_lowerlim))]
-            # Define mean value as nominal slit distance
-            if len(__diff) == 0:
-                continue
-            slit_dist = np.mean(__diff)
-
-            # Conditions for idenfitying the ref slit
-            # Slit distance somewhere between 2 slits:
-            # 2 slit distances (defined above) +/- ref_range
-            ref_hi_cond = ((self._ref_edges + 2) *
-                           slit_dist * (1 + self._ref_range))
-            ref_lo_cond = ((self._ref_edges + 1) *
-                           slit_dist * (1 - self._ref_range))
-            # Find the reference slit locations (indexes)
-            _ref_idx = np.argwhere(np.logical_and(
-                _diff < ref_hi_cond, _diff > ref_lo_cond)).flatten()
-            if len(_ref_idx) != 1:
-                continue
-            self._ref_indexes.append(_ref_idx[0] + offset)
-            offset += len(diff_split[i])
+        # Function to find the mean off an array without outliers
+        def _mean(arr):
+            high = np.percentile(arr, 90)
+            low = np.percentile(arr, 10)
+            return np.mean(arr[(arr < high) & (arr > low)])
+
+        # Generate self._encd_diff
+        self._generate_sub_data(ref_clk=False)
+
+        # Find the instantaneous average encoder clk spacing between datapoints
+        rot_spacing = np.arange(0, len(self._encd_diff), self._num_edges - 2)
+        diff_split = np.split(self._encd_diff, rot_spacing[1:])
+        slit_dist = np.concatenate([np.ones(len(_diff)) * _mean(_diff) for _diff in diff_split])
+
+        # Generate thresholds to distinguish references and glitched references
+        ref_hi_cond = (self._ref_edges + 1) * slit_dist * (1 + self._ref_range)
+        ref_lo_cond = (self._ref_edges + 1) * slit_dist * (1 - self._ref_range)
+        ref_gl_cond = slit_dist * (1 + self._ref_range)
+
+        # Normal references
+        ref_ind = np.where((self._encd_diff > ref_lo_cond) & (self._encd_diff < ref_hi_cond))[0]
+        # Potential glitched references
+        ref_ind_glitch = np.where((self._encd_diff < ref_lo_cond) & (self._encd_diff > ref_gl_cond))[0]
+
+        # Find and add glitched references to normal references
+        used_ind_pairs = []
+        for ind in ref_ind_glitch:
+            # Check that there is sufficient space before and after the glitched reference
+            before = min(ref_ind, key=lambda x:abs(x - ind) + 1e8*(1 + np.sign(x - ind)))
+            after = min(ref_ind, key=lambda x:abs(x - ind) + 1e8*(1 - np.sign(x - ind)))
```

Comment:
Thank you so much Bryce! I will test this on many observations before I merge.

By the way, I found that it is taking significantly long time for these two lines.
Replacing with something like following will improve the computation time a lot.

```
def testp(x, ind):
    return abs(x - ind) + 1e8*(1 - np.sign(x + ind))
def testm(x, ind):
    return abs(x - ind) + 1e8*(1 - np.sign(x - ind))

for ind in ref_ind_glitch:
    # Check that there is sufficient space before and after the glitched reference
    before = ref_ind[np.argmin(testp(ref_ind, ind))]
    after = ref_ind[np.argmin(testm(ref_ind, ind))]
```

Reply:
These lines were making the typical processing time of 1hour observation data from ~30 seconds to 10 minutes.

I finally identified why it takes very long time on these lines for SO data.
It's because conditions of glitched reference is too strict and half of the all the data points were counted as the potential glitched references. I think this happens because SO encoder data is less uniform compared to pb2b data.

Reply:
I think this is fixed by https://github.com/simonsobs/sotodlib/pull/1021/commits/a12bf0422ae618f8102edc216f2674784ac1e3e0 I will work on testing this again.